RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+      "$id": "#/properties/paths",
+      "title": "Paths property",
+      "description": "Array of file paths to also ignore by the PDK, for example, when building a module",
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+    },
+  },
+  "definitions": {}
+}
+```
+
+**Note** that the `required` setting is not documented, or has a schema, as it's internal to the template.
+
+`.pdkignore.erb`
+
+Same as [V1 template](https://github.com/puppetlabs/pdk-templates/blob/90e5efab878facd18f838755d9e95b226b1e0776/moduleroot/.pdkignore.erb).
+
+#### An opt-in template for CI configuration
+
+> TODO
+
+#### A template only for new object
+
+
+### Changes to `.sync.yml`
+
+Each module may contain a `.sync.yml` file which contains the module specific settings when rendering the templates. For the new templates, the .sync.yml will have additional settings, which will also be backwards compatible with the V1 templating engine.
+
+* Root `pdk_template` element
+
+  The new `pdk_template` element will hold all of the data for templating engine. Due to being backwards compatible the verbose name of `pdk_template` is used instead of the generic name `template`, so that it would probably not conflict with any known files.  For example, Puppet modules have a `templates` directory.
+
+  * `version` element (Integer)
+
+  Which version of the templating engine this module uses. If this is missing (or the root `pdk_template` element is missing) it is assumed this is Version 1 template. If the template specifies a version that the templating engine does not support it is expected the engine will raise a terminating error.
+
+  Note that the type here is an integer. Although we could use a semver (Semantic Versioning) based string here, that would make parsing difficult.
+
+  * `templates` element (Array of string). List of templates (The template directory name) to be applied **IN ORDER** (Order is important.  See [Resolution Order](#resolution_order))
+
+Example V1 `.sync.yml`
+
+``` yaml
+---
+Gemfile:
+  # Gemfile template customisations
+
+.rubocop.yml:
+  # Rubocop template customisations
+
+# ... etc
+```
+
+Example `.sync.yml` with V2 settings
+
+``` yaml
+---
+pdk_template:
+  version: 2
+  templates:
+    - appveyor_ci
+    - travis_ci
+    - rubocop
+    - litmus_tests
+
+Gemfile:
+  # Gemfile template customisations
+
+.rubocop.yml:
+  # Rubocop template customisations
+
+# ... etc
+```
+
+### Resolution order
+
+> TODO: What happens with conflicting settings files.  Order is important, first one wins.
+> For `always_apply` modules they are last, but in alphabetical order (Needs some kind of deterministic order)
+
+## Unresolved Questions
+
+> TODO:
+> Having discrete settings may cause duplicate code because you can't share across templates.  But that's better
+> spaghetti dependencies and difficult to diagnose. (DRY isn't always best)
+
+## Future Considerations
+
+> TODO:
+> Rake tasks to auto generate documentation
```

Comment:
Does this fit into puppet-strings? Might be too far-fetched, and overcomplicated.

Reply:
Probably not the correct place. PDK maybe ...